stdlib/public/core/CollectionAlgorithms.swift

```diff
@@ -209,6 +209,84 @@ extension BidirectionalCollection where Element: Equatable {
   }
 }
 
+//===----------------------------------------------------------------------===//
+// indices(where:) / indices(of:)
+//===----------------------------------------------------------------------===//
+
+#if !$Embedded
+extension Collection {
+  /// Returns the indices of all the elements that match the given predicate.
+  ///
+  /// For example, you can use this method to find all the places that a
+  /// vowel occurs in a string.
+  ///
+  ///     let str = "Fresh cheese in a breeze"
+  ///     let vowels: Set<Character> = ["a", "e", "i", "o", "u"]
+  ///     let allTheVowels = str.indices(where: { vowels.contains($0) })
+  ///     // str[allTheVowels].count == 9
+  ///
+  /// - Parameter predicate: A closure that takes an element as its argument
+  ///   and returns a Boolean value that indicates whether the passed element
+  ///   represents a match.
+  /// - Returns: A set of the indices of the elements for which `predicate`
+  ///   returns `true`.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  @available(SwiftStdlib 5.11, *)
+  public func indices(where predicate: (Element) throws -> Bool) rethrows
+    -> RangeSet<Index>
+  {
+    if isEmpty { return RangeSet() }
+
+    var result: [Range<Index>] = []
+    var i = startIndex
```

Comment:
It seems to me that we ought to be able to write the search loop considerably more clearly by taking advantage of existing API, something like:
```
     var end = startIndex
     while end != endIndex {
        if let begin = try self[end...].firstIndex(where: predicate) {
            end = try self[begin...].prefix(while: predicate).endIndex
            result.append(begin ..< end)
            formIndex(after: &end)
        }
     }
```

Reply:
Good point, will update